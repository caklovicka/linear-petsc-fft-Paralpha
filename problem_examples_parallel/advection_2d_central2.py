import numpy as np
from scipy import sparse
from petsc4py import PETSc
from core_parallel.linear_paralpha import LinearParalpha

"""
advection eq. in 2d
u_t + c_x * u_x + c_y * u_y = f
"""


class Advection(LinearParalpha):

    # user defined, just for this class
    c = [1, 1]
    X_left = 0
    X_right = 1
    Y_left = 0
    Y_right = 1
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 2)[1:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 2)[1:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / (self.spatial_points[0] + 1))
        self.dx.append((self.Y_right - self.Y_left) / (self.spatial_points[1] + 1))

        # x and size_global_A have to be filled before super().setup()
        self.x = np.meshgrid(self.xx, self.yy)
        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        for i in range(self.row_beg, self.row_end, 1):
            # right x diagonal
            if (i+1) % self.spatial_points[0] != 0:
                row.append(i)
                col.append(i+1)
                data.append(-self.c[0] / (2 * self.dx[0]))
            # left x diagonal
            if i % self.spatial_points[0] != 0:
                row.append(i)
                col.append(i - 1)
                data.append(self.c[0] / (2 * self.dx[0]))

            # left y diagonal
            if i >= self.spatial_points[0]:
                row.append(i)
                col.append(i-self.spatial_points[0])
                data.append(self.c[1] / (2 * self.dx[1]))
            # right y diagonal
            if i + self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i+self.spatial_points[0])
                data.append(-self.c[1] / (2 * self.dx[1]))

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):

        bb = self.rhs(t, self.x).flatten()
        bb[:self.spatial_points[0]] += self.c[1] / (2 * self.dx[1]) * self.e(t, self.xx)
        bb[-self.spatial_points[0]:] -= self.c[1] / (2 * self.dx[1]) * self.d(t, self.xx)
        bb[::self.spatial_points[0]] += self.c[0] / (2 * self.dx[0]) * self.a(t, self.yy)
        bb[self.spatial_points[0]-1::self.spatial_points[0]] -= self.c[0] / (2 * self.dx[0]) * self.b(t, self.yy)

        return bb[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        return np.sin(2 * np.pi * (z[0] - self.c[0] * t)) * np.sin(2 * np.pi * (z[1] - self.c[1] * t))

    # user defined
    def a(self, t, y):
        return self.u_exact(t, (self.X_left, y))

    # user defined
    def b(self, t, y):
        return self.u_exact(t, (self.X_right, y))

    # user defined
    def e(self, t, x):
        return self.u_exact(t, (x, self.Y_left))

    # user defined
    def d(self, t, x):
        return self.u_exact(t, (x, self.Y_right))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    @staticmethod
    def rhs(t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, atol=tol, max_it=self.global_size_A)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it







