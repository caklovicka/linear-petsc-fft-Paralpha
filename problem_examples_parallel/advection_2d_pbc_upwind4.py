import numpy as np
from scipy import sparse
from petsc4py import PETSc
from core_parallel.linear_paralpha import LinearParalpha

"""
advection1 eq. in 2d
u_t + c_x * u_x + c_y * u_y = f
"""


class Advection(LinearParalpha):

    # user defined, just for this class
    c = [1, 1]
    X_left = 0
    X_right = 1
    Y_left = 0
    Y_right = 1
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 1)[:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 1)[:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / self.spatial_points[0])
        self.dx.append((self.Y_right - self.Y_left) / self.spatial_points[1])

        # x and size_global_A have to be filled before super().setup()
        self.x = np.meshgrid(self.xx, self.yy)
        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = -self.c[0] / (60 * self.dx[0])
        cy = -self.c[1] / (60 * self.dx[1])
        for i in range(self.row_beg, self.row_end, 1):

            row.append(i)
            col.append((i - 3) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-5 * cx)

            row.append(i)
            col.append((i - 2) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(30 * cx)

            row.append(i)
            col.append((i - 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-90 * cx)

            row.append(i)
            col.append(i % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(50 * cx + 50 * cy)

            row.append(i)
            col.append((i + 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(15 * cx)

            row.append(i)
            col.append((i - 3 * self.spatial_points[0]) % self.global_size_A)
            data.append(-5 * cy)

            row.append(i)
            col.append((i - 2 * self.spatial_points[0]) % self.global_size_A)
            data.append(30 * cy)

            row.append(i)
            col.append((i - self.spatial_points[0]) % self.global_size_A)
            data.append(-90 * cy)

            row.append(i)
            col.append((i + self.spatial_points[0]) % self.global_size_A)
            data.append(15 * cy)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))
        del data, row, col

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        return np.sin(2 * np.pi * (z[0] - self.c[0] * t)) * np.sin(2 * np.pi * (z[1] - self.c[1] * t))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    @staticmethod
    def rhs(t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it







