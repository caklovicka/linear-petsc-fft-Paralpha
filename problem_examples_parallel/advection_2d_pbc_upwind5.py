import numpy as np
from scipy import sparse
from petsc4py import PETSc
from mpi4py import MPI
from core_parallel.linear_paralpha import LinearParalpha

"""
advection1 eq. in 2d
u_t + c_x * u_x + c_y * u_y = f
"""


class Advection(LinearParalpha):

    # user defined, just for this class
    c = [1, 1]
    X_left = 0
    X_right = 1
    Y_left = 0
    Y_right = 1
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 1)[:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 1)[:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / self.spatial_points[0])
        self.dx.append((self.Y_right - self.Y_left) / self.spatial_points[1])

        # x and size_global_A have to be filled before super().setup()
        self.x = np.meshgrid(self.xx, self.yy)
        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = -self.c[0] / (60 * self.dx[0])
        cy = -self.c[1] / (60 * self.dx[1])
        for i in range(self.row_beg, self.row_end, 1):
            row.append(i)
            col.append((i - 4) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(3 * cx)

            row.append(i)
            col.append((i - 3) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-20 * cx)

            row.append(i)
            col.append((i - 2) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(60 * cx)

            row.append(i)
            col.append((i - 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(-120 * cx)

            row.append(i)
            col.append(i % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(65 * cx + 65 * cy)

            row.append(i)
            col.append((i + 1) % self.spatial_points[0] + (i // self.spatial_points[0]) * self.spatial_points[0])
            data.append(12 * cx)

            row.append(i)
            col.append((i - 4 * self.spatial_points[0]) % self.global_size_A)
            data.append(3 * cy)

            row.append(i)
            col.append((i - 3 * self.spatial_points[0]) % self.global_size_A)
            data.append(-20 * cy)

            row.append(i)
            col.append((i - 2 * self.spatial_points[0]) % self.global_size_A)
            data.append(60 * cy)

            row.append(i)
            col.append((i - self.spatial_points[0]) % self.global_size_A)
            data.append(-120 * cy)

            row.append(i)
            col.append((i + self.spatial_points[0]) % self.global_size_A)
            data.append(12 * cy)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        # ---- POSTSETUP <end> ----
        return

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        return np.sin(2 * np.pi * (z[0] - self.c[0] * t)) * np.sin(2 * np.pi * (z[1] - self.c[1] * t))

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    @staticmethod
    def rhs(t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x.real, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        print('tu sam')
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create(comm=self.comm_matrix)
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it