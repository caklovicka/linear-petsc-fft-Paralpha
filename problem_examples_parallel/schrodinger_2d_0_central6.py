import numpy as np
from scipy import sparse
from core_parallel.linear_paralpha import LinearParalpha
import scipy as sc
from petsc4py import PETSc

"""
schrodinger in 2d, 2nd order in space
u_t = c ( u_xx + u_yy )
"""


class Schrodinger(LinearParalpha):

    # user defined, just for this class
    c = 1j
    sigma = 1
    p = [1, 1]
    X_left = -10
    X_right = 10
    Y_left = -10
    Y_right = 10
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0] + 2)[1:-1]
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1] + 2)[1:-1]

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / (self.spatial_points[0] + 1))
        self.dx.append((self.Y_right - self.Y_left) / (self.spatial_points[1] + 1))

        # x and size_global_A have to be filled before super().setup()

        self.x = np.meshgrid(self.xx, self.yy)
        del self.xx, self.yy

        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = self.c / self.dx[0]**2
        cy = self.c / self.dx[1]**2
        for i in range(self.row_beg, self.row_end, 1):
            # x part
            # i - 3
            if i % self.spatial_points[0] - 3 >= 0:
                row.append(i)
                col.append(i - 3)
                data.append(cx / 90)

            # i - 2
            if i % self.spatial_points[0] - 2 >= 0:
                row.append(i)
                col.append(i - 2)
                data.append(-3 * cx / 20)

            # i - 1
            if i % self.spatial_points[0] - 1 >= 0:
                row.append(i)
                col.append(i - 1)
                data.append(3 * cx / 2)

            # i
            row.append(i)
            col.append(i)
            data.append(-49 * cx / 18 - 49 * cy / 18)

            # i + 1
            if i % self.spatial_points[0] + 1 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 1)
                data.append(3 * cx / 2)

            # i + 2
            if i % self.spatial_points[0] + 2 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 2)
                data.append(-3 * cx / 20)

            # i + 3
            if i % self.spatial_points[0] + 3 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 3)
                data.append(cx / 90)

            # y part
            # i - 3
            if i - 3 * self.spatial_points[0] >= 0:
                row.append(i)
                col.append(i - 3 * self.spatial_points[0])
                data.append(cy / 90)

            # i - 2
            if i - 2 * self.spatial_points[0] >= 0:
                row.append(i)
                col.append(i - 2 * self.spatial_points[0])
                data.append(-3 * cy / 20)

            # i - 1
            if i - self.spatial_points[0] >= 0:
                row.append(i)
                col.append(i - self.spatial_points[0])
                data.append(3 * cy / 2)

            # i + 1
            if i + self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + self.spatial_points[0])
                data.append(3 * cy / 2)

            # i + 2
            if i + 2 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 2 * self.spatial_points[0])
                data.append(-3 * cy / 20)

            # i + 3
            if i + 3 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 3 * self.spatial_points[0])
                data.append(cy / 90)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        del data, row, col

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        a = self.sigma + 2j * t
        const = self.sigma / a * np.exp(self.sigma/2 * (self.p[0]**2 + self.p[1]**2) * (self.sigma/a - 1))
        exp1 = np.exp(1j * self.sigma / a * (self.p[0] * z[0] + self.p[1] * z[1]))
        exp2 = np.exp(-1/(2 * a) * (z[0]**2 + z[1]**2))
        return const * exp1 * exp2

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    def rhs(self, t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it
