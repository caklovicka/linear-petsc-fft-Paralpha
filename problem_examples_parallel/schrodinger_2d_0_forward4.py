import numpy as np
from scipy import sparse
from core_parallel.linear_paralpha import LinearParalpha
import scipy as sc
from petsc4py import PETSc

"""
schrodinger in 2d, 2nd order in space
u_t = c ( u_xx + u_yy )
"""


class Schrodinger(LinearParalpha):

    # user defined, just for this class
    c = 1j
    sigma = 1
    p = [1, 1]
    X_left = -10
    X_right = 10
    Y_left = -10
    Y_right = 10
    xx = None
    yy = None

    # PROCESSORS NEEDED = proc_row * proc_col

    def __init__(self):

        LinearParalpha.__init__(self)

    def setup(self):

        # ---- PRESETUP ----

        self.xx = np.linspace(self.X_left, self.X_right, self.spatial_points[0])
        self.yy = np.linspace(self.Y_left, self.Y_right, self.spatial_points[1])

        self.dx = []
        self.dx.append((self.X_right - self.X_left) / (self.spatial_points[0] - 1))
        self.dx.append((self.Y_right - self.Y_left) / (self.spatial_points[1] - 1))

        # x and size_global_A have to be filled before super().setup()

        self.x = np.meshgrid(self.xx, self.yy)
        del self.xx, self.yy

        self.global_size_A = 1
        for n in self.spatial_points:
            self.global_size_A *= n

        # ---- PRESETUP <end> ----

        # do not delete this, this builds communicators needed for your matrix Apar
        super().setup()

        # ---- POSTSETUP ----

        # user defined, parallel matrix, rows are divided between processors in group 'comm_matrix'
        # rows_beg and rows_end define the chunk of matrix Apar in this communicator

        # user defined matrix Apar, sparse form recommended
        row = list()
        col = list()
        data = list()
        cx = self.c / self.dx[0]**2
        cy = self.c / self.dx[1]**2

        # 15/4	−77/6	107/6	−13	   61/12	 −5/6
        for i in range(self.row_beg, self.row_end, 1):
            # x part
            row.append(i)
            col.append(i)
            data.append(15/4 * cx + 15/4 * cy)

            if i % self.spatial_points[0] + 1 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 1)
                data.append(-77/6 * cx)

            if i % self.spatial_points[0] + 2 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 2)
                data.append(107/6 * cx)

            if i % self.spatial_points[0] + 3 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 3)
                data.append(-13 * cx)

            if i % self.spatial_points[0] + 4 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 4)
                data.append(61/12 * cx)

            if i % self.spatial_points[0] + 5 < self.spatial_points[0]:
                row.append(i)
                col.append(i + 5)
                data.append(-5/6 * cx)

            # y part
            if i + self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + self.spatial_points[0])
                data.append(-77/6 * cy)

            if i + 2 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 2 * self.spatial_points[0])
                data.append(107/6 * cy)

            if i + 3 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 3 * self.spatial_points[0])
                data.append(-13 * cy)

            if i + 4 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 4 * self.spatial_points[0])
                data.append(61/12 * cy)

            if i + 5 * self.spatial_points[0] < self.global_size_A:
                row.append(i)
                col.append(i + 5 * self.spatial_points[0])
                data.append(-5/6 * cy)

        data = np.array(data)
        row = np.array(row) - self.row_beg
        col = np.array(col)
        self.Apar = sparse.csr_matrix((data, (row, col)), shape=(self.row_end - self.row_beg, self.global_size_A))

        del data, row, col

        # ---- POSTSETUP <end> ----

    # user defined
    def bpar(self, t):
        return self.rhs(t, self.x).flatten()[self.row_beg:self.row_end]

    # user defined
    def u_exact(self, t, z):
        a = self.sigma + 2j * t
        const = self.sigma / a * np.exp(self.sigma/2 * (self.p[0]**2 + self.p[1]**2) * (self.sigma/a - 1))
        exp1 = np.exp(1j * self.sigma / a * (self.p[0] * z[0] + self.p[1] * z[1]))
        exp2 = np.exp(-1/(2 * a) * (z[0]**2 + z[1]**2))
        return const * exp1 * exp2

    # DELETE
    def u_t(self, t, z):
        a = self.sigma + 2j * t
        exp_const1 = self.sigma/2 * (self.p[0]**2 + self.p[1]**2) * (self.sigma/a - 1)
        exp_const2 = 1j * self.sigma / a * (self.p[0] * z[0] + self.p[1] * z[1])
        exp_const3 = -1/(2 * a) * (z[0]**2 + z[1]**2)
        exp = np.exp(exp_const1 + exp_const2 + exp_const3)

        sum1 = -2j * self.sigma / a**2 * exp
        sum2 = self.sigma/a * exp * (-self.sigma**2 * 2j / a**2 + (z[0]**2 + z[1]**2) * 1j / a**2 - 1j * self.sigma * 2j/a**2)

        return sum1 + sum2

    # user defined
    def u_initial(self, z):
        return self.u_exact(self.T_start, z)

    # user defined
    def rhs(self, t, z):
        return 0 * z[0] * z[1]

    @staticmethod
    def norm(x):
        return np.linalg.norm(x, np.inf)

    # petsc solver on comm_matrix
    def linear_solver(self, M_loc, m_loc, m0_loc, tol):
        m = PETSc.Vec()
        m.createWithArray(array=m_loc, comm=self.comm_matrix)
        m0 = PETSc.Vec()
        m0.createWithArray(array=m0_loc, comm=self.comm_matrix)
        M = PETSc.Mat()
        csr = (M_loc.indptr, M_loc.indices, M_loc.data)
        M.createAIJ(size=(self.global_size_A, self.global_size_A), csr=csr, comm=self.comm_matrix)

        ksp = PETSc.KSP()
        ksp.create()
        ksp.setType('gmres')
        ksp.setFromOptions()
        ksp.setTolerances(rtol=tol, max_it=self.smaxiter)
        pc = ksp.getPC()
        pc.setType('none')
        ksp.setOperators(M)
        ksp.setInitialGuessNonzero(True)
        ksp.solve(m, m0)
        sol = m0.getArray()
        it = ksp.getIterationNumber()

        m.destroy()
        m0.destroy()
        ksp.destroy()
        M.destroy()

        return sol, it
