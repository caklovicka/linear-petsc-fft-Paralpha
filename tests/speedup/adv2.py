import os
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"] = "1"

import sys
sys.path.append('../..')
import numpy as np
from mpi4py import MPI
from problem_examples_parallel.advection_2d_pbc_upwind3 import Advection as Adv3

prob = Adv3()
N = 800
prob.spatial_points = [N, N]
prob.tol = 1e-9
prob.proc_col = 1
prob.time_points = 2
prob.optimal_alphas = False
prob.alphas = [1e-8]
prob.T_start = 0
prob.T_end = 0.00064
prob.solver = 'custom'
prob.maxiter = 5
prob.smaxiter = 1000
prob.time_intervals = 64
prob.proc_row = 64

prob.stol = 1e-11
if prob.rolling < 64:
    prob.stol = 1e-13
    prob.tol = 1e-9 / prob.rolling

prob.m0 = 10 * (prob.T_end - prob.T_start)/prob.rolling

prob.setup()
prob.solve()
prob.summary(details=True)

if prob.rank == prob.size - 1: # prob.size_subcol_seq:
    exact = prob.u_exact(prob.T_end, prob.x).flatten()[prob.row_beg:prob.row_end]
    approx = prob.u_last_loc.flatten()
    d = exact - approx
    d = d.flatten()
    err_abs = np.linalg.norm(d, np.inf)
    # err_abs_root = prob.comm_subcol_seq.reduce(err_abs, op=MPI.MAX, root=prob.size_subcol_seq - 1)
    # if prob.rank == prob.size - 1:
    print('abs err = {}'.format(err_abs))
