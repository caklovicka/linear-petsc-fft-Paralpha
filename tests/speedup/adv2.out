on 63,  abs, rel err inf norm [from paralpha] = 4.023234292382938e-07, 4.023293393948375e-07, iter = 0, rolling = 0
on 63,  abs, rel err inf norm [from paralpha] = 3.2925295823105216e-10, 3.292577949777693e-10, iter = 1, rolling = 0
-----------------------< summary >-----------------------
solving on T1 = 0, T2 = 0.00064
no. of spatial points = [800, 800]
dx = [0.00125, 0.00125]
no. of time points on an interval = 2
no. of time intervals = 64
no. of alphas = 1, [0.0001]
rolling intervals = 1
dt (of one SDC interval) = 1e-05
processors for spatial parallelization for solving (I - Q x A) are 1
processors for time interval parallelization are 64
maxiter = 10
output document = None
tol = 1e-09
last error = NotImplemented
residuals = [[6.283185560052029e-05, 4.0216005448012737e-07, 4.022016053539801e-11]]
consecutive errors= [[inf, 0.00402160054466021, 4.02203227248938e-07]]
iterations of Paralpha = [2.]
max iterations of Paralpha = 2.0
algorithm time = 3.76745 s
communication time = 0.66308 s
inner solver = custom
system_time_max = [1.2448985940000057, 1.3007045669996842]
system_time_min = [1.1869509070002096, 1.1847698749998017]
solver_its_max = [[2, 4]]
solver_its_min = [[2, 2]]
inner solver tols = [1e-13, 1e-13]
inner solver maxiter = 500
-----------------------< end summary >-----------------------
abs err = 0.004021198223405575
on 63,  abs, rel err inf norm [from paralpha] = 3.257770698613635e-10, 3.2578185554706136e-10, iter = 0, rolling = 0
-----------------------< summary >-----------------------
solving on T1 = 0, T2 = 0.00064
no. of spatial points = [800, 800]
dx = [0.00125, 0.00125]
no. of time points on an interval = 2
no. of time intervals = 64
no. of alphas = 1, [1e-08]
rolling intervals = 1
dt (of one SDC interval) = 1e-05
processors for spatial parallelization for solving (I - Q x A) are 1
processors for time interval parallelization are 64
maxiter = 5
output document = None
tol = 1e-09
last error = NotImplemented
residuals = [[6.283185560052029e-05, 4.0212055907318245e-11]]
consecutive errors= [[inf, 0.004021198375211929]]
iterations of Paralpha = [1.]
max iterations of Paralpha = 1.0
algorithm time = 1.95702 s
communication time = 0.34912 s
inner solver = custom
system_time_max = [1.269568563000007]
system_time_min = [1.185645582999996]
solver_its_max = [[3]]
solver_its_min = [[2]]
inner solver tols = [1e-13]
inner solver maxiter = 1000
-----------------------< end summary >-----------------------
abs err = 0.004021198223405575
